File: bukkit/elysium-chat-bukkit/src/main/java/com/seventh_root/elysium/chat/bukkit/listener/AsyncPlayerChatListener.java
package com.seventh_root.elysium.chat.bukkit.listener;

import com.seventh_root.elysium.api.player.ElysiumPlayer;
import com.seventh_root.elysium.chat.bukkit.ElysiumChatBukkit;
import com.seventh_root.elysium.chat.bukkit.chatchannel.BukkitChatChannel;
import com.seventh_root.elysium.chat.bukkit.chatchannel.BukkitChatChannelProvider;
import com.seventh_root.elysium.chat.bukkit.context.BukkitChatMessageContext;
import com.seventh_root.elysium.players.bukkit.BukkitPlayer;
import com.seventh_root.elysium.players.bukkit.BukkitPlayerProvider;
import org.bukkit.ChatColor;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class AsyncPlayerChatListener implements Listener {

    private final ElysiumChatBukkit plugin;

    public AsyncPlayerChatListener(ElysiumChatBukkit plugin) {
        this.plugin = plugin;
    }

    @EventHandler
    public void onAsyncPlayerChat(AsyncPlayerChatEvent event) {
        event.setCancelled(true);
        Player bukkitPlayer = event.getPlayer();
        BukkitPlayerProvider playerProvider = plugin.getCore().getServiceManager().getServiceProvider(BukkitPlayerProvider.class);
        BukkitChatChannelProvider chatChannelProvider = plugin.getCore().getServiceManager().getServiceProvider(BukkitChatChannelProvider.class);
        BukkitPlayer player = playerProvider.getPlayer(bukkitPlayer);
        BukkitChatChannel channel = chatChannelProvider.getPlayerChannel(player);
        String message = event.getMessage();
        for (BukkitChatChannel otherChannel : chatChannelProvider.getChatChannels()) {
            if (!otherChannel.getMatchPattern().isEmpty() && message.matches(otherChannel.getMatchPattern())) {
                channel = otherChannel;
                Pattern pattern = Pattern.compile(otherChannel.getMatchPattern());
                Matcher matcher = pattern.matcher(message);
                if (matcher.matches()) {
                    if (matcher.groupCount() > 0) {
                        message = matcher.group(1);
                    }
                }
                if (!channel.getListeners().contains(player)) {
                    channel.addListener(player);
                    chatChannelProvider.updateChatChannel(channel);
                }
            }
        }
        if (channel != null) {
            for (ElysiumPlayer listener : channel.getListeners()) {
                if (listener instanceof BukkitPlayer) {
                    BukkitPlayer bukkitListener = (BukkitPlayer) listener;
                    OfflinePlayer bukkitOfflinePlayer = bukkitListener.getBukkitPlayer();
                    if (bukkitOfflinePlayer.isOnline()) {
                        if (channel.getRadius() <= 0
                                || bukkitPlayer.getLocation().distanceSquared(bukkitOfflinePlayer.getPlayer().getLocation()) <= channel.getRadius() * channel.getRadius()) {
                            String processedMessage = channel.processMessage(message, new BukkitChatMessageContext(channel, player, listener));
                            if (processedMessage != null) {
                                bukkitOfflinePlayer.getPlayer().sendMessage(processedMessage);
                            }
                        }
                    }
                }
            }
        } else {
            event.getPlayer().sendMessage(ChatColor.translateAlternateColorCodes('&', plugin.getConfig().getString("messages.no-chat-channel")));
        }
    }

}

File: bukkit/elysium-chat-bukkit/src/main/kotlin/com/seventh_root/elysium/chat/bukkit/listener/AsyncPlayerChatListener.kt
/*
 * Copyright 2016 Ross Binden
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.seventh_root.elysium.chat.bukkit.listener

import com.seventh_root.elysium.chat.bukkit.ElysiumChatBukkit
import com.seventh_root.elysium.chat.bukkit.chatchannel.ElysiumChatChannelProvider
import com.seventh_root.elysium.players.bukkit.player.ElysiumPlayerProvider
import org.bukkit.ChatColor
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent

class AsyncPlayerChatListener(private val plugin: ElysiumChatBukkit): Listener {

    @EventHandler
    fun onAsyncPlayerChat(event: AsyncPlayerChatEvent) {
        event.isCancelled = true
        val chatChannelProvider = plugin.core.serviceManager.getServiceProvider(ElysiumChatChannelProvider::class)
        val playerProvider = plugin.core.serviceManager.getServiceProvider(ElysiumPlayerProvider::class)
        val player = playerProvider.getPlayer(event.player)
        val chatChannel = chatChannelProvider.getPlayerChannel(player)
        if (chatChannel != null) {
            chatChannel.sendMessage(player, event.message)
        } else {
            event.player.sendMessage(ChatColor.translateAlternateColorCodes('&', plugin.config.getString("messages.no-chat-channel")))
        }
    }

}
